Add abode config entries and device registry

shred86 opened the pull request:
## Breaking Changes:

This PR removes the following configuration variables: `name`, `exclude` and `lights`, which were all previously optional. Existing users of the abode integration that use these configuration variables will have to remove them from the configuration.yaml file. Entities that users wish to disable can be done from the Entity Registry in the Configuration UI.

## Description:

Adds config entry and entity registry support for the Abode component. This implementation will create a config entry if a user has added the abode component configured via the configuration.yaml file. The entity registry unique ID is based on `device_id` from Abode (no devices in Abode should ever have the same device_id).

Tested on my Abode setup which includes sensors and lights.

## Docs PR:
https://github.com/home-assistant/home-assistant.io/pull/10467

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

shred86 committed:
Adds support for config entries and device registry

shred86 committed:
Fixing string formatting for logger

shred86 commented:
Hey @frenck, what type of documentation would I need to create? The setup configuration settings if using the .yaml file hasn’t changed. I looked at a few other components that support config entires and didn’t really see anything I needed to add. Let me know and I’ll definitely get it taken care of! Thanks.

frenck commented:
@shred86 The documentation page needs a frontmatter option: `ha_config_flow` which is set to `true`, furthermore, it would be helpful to explain to a user how to activate/add the integration using a config flow.

As for this PR, please add unit tests. Config flows require unit tests.

shred86 commented:
> @shred86 The documentation page needs a frontmatter option: `ha_config_flow` which is set to `true`, furthermore, it would be helpful to explain to a user how to activate/add the integration using a config flow.
> 
> As for this PR, please add unit tests. Config flows require unit tests.

Ah okay, thanks. Just created a pull request with the updated documentation (https://github.com/home-assistant/home-assistant.io/pull/10386). I’ll try to add unit tests soon. Unfortunately I’m fairly new to Python and will be fairly busy for the next two weeks.

shred86 commented:
Unfortunately I'm struggling with unit tests. If there's any guidance/documentation specific to writing unit tests for Home Assistant, please let me know. The basics of using the unittest module in python makes sense, but I'm not having much luck in understanding how they're written for Home Assistant, or more specifically how to execute async unit tests. I'll keep at it to try and figure it out, but any assistance would be appreciated.

MartinHjelmare commented:
Read the pytest docs. Read a blog on unit testing eg this one:
https://realpython.com/python-testing/

We want all new tests to be standalone pytest test functions.

See existing tests eg:
https://github.com/home-assistant/home-assistant/blob/dev/tests/components/hue/test_config_flow.py

shred86 committed:
Add unit test for abode config flow

shred86 commented on tests/components/abode/test_config_flow.py:
```diff
@@ -0,0 +1,80 @@
+    flow = config_flow.AbodeFlowHandler()
+    flow.hass = hass
+
+    with patch("abodepy.Abode", side_effect=AbodeAuthenticationException("errors")):
```
Didn't quite understand this but I had to add an argument to `AbodeAuthenticationException` for it to work. Does not make sense to me why it requires an argument.

MartinHjelmare replied:
That exception requires a tuple as argument:
https://github.com/MisterWil/abodepy/blob/6f84bb428fd1da98855f55083cd427bebbcc57ae/abodepy/exceptions.py#L4-L17

https://github.com/MisterWil/abodepy/blob/61bb200ed21153a6d03e68044e9a729778f31d0a/abodepy/__init__.py#L139-L140

shred86 replied:
I'm an idiot - that is very obvious now, lol. If I'm understanding this correctly, it really doesn't matter what argument is passed since really all I need is the `AbodeAuthenticationException` to assert to test that the config flow is functioning properly. Doesn't look like abodepy is currently doing anything with the arguments that are passed anyways.

MartinHjelmare replied:
The argument must support access by index, to avoid raising an error when creating the exception. After that it may not matter. I guess it depends on what we do when excepting the exception.

shred86 commented:
Added unit test for config flows. I used the SimpliSafe component as a reference. I think I covered all the test cases needed.

shred86 committed:
Fix for lights, only allow one config, add ability to unload entry

shred86 committed:
Fix for subscribing to hass_events on adding abode component

shred86 reviewed (commented):
Just pushed a fix for this issue. Previously, when `setup_hass_events` was being called, it would execute `hass.bus.listen_once(EVENT_HOMEASSISTANT_START, startup)`. This is why it was only working after a reboot since it was waiting for the `EVENT_HOMEASSISTANT_STARTUP` event to fire. I couldn't think of any reason not to simply call the `startup` function when `setup_hass_events` is called (except for the case where polling is enabled), so that's what the changes reflect in addition to making the code async. The function `setup_hass_events` should only be called during initial HA start or when the abode component is added, so there shouldn't ever be a situation where it's called twice.

MartinHjelmare commented on homeassistant/components/abode/alarm_control_panel.py:
```diff
@@ -9,22 +9,24 @@
+    data = hass.data[DOMAIN]
 
-    add_entities(alarm_devices)
+    async_add_devices([AbodeAlarm(data, data.abode.get_alarm(), data.name)], True)
```
Does `data.abode.get_alarm()` do I/O?

shred86 replied:
It appears to do a `get` request. From abodepy:

`response = self.send_request("get", CONST.DEVICES_URL)`

MartinHjelmare replied:
Then we need to schedule it on the executor thread pool. Use `await hass.async_add_executor_job(...)`.

MartinHjelmare commented on homeassistant/components/abode/camera.py:
```diff
@@ -7,19 +7,25 @@
+    data = hass.data[DOMAIN]
 
     devices = []
     for device in data.abode.get_devices(generic_type=CONST.TYPE_CAMERA):
```
I/O?

shred86 replied:
I don't believe this one does since a `generic_type` is passed.

https://github.com/MisterWil/abodepy/blob/6f84bb428fd1da98855f55083cd427bebbcc57ae/abodepy/__init__.py#L241

MartinHjelmare replied:
Do we know that devices are already cached?
https://github.com/MisterWil/abodepy/blob/6f84bb428fd1da98855f55083cd427bebbcc57ae/abodepy/__init__.py#L192

shred86 replied:
I believe so because when `abodepy.Abode` is [called](https://github.com/home-assistant/home-assistant/blob/e005f6f23a3ff9ff052afd317891da206ea618e0/homeassistant/components/abode/__init__.py#L107) during the setup of `AbodeSystem`, `get_devices` is set to `True` which should populate `_devices`.

MartinHjelmare commented on homeassistant/components/abode/config_flow.py:
```diff
@@ -0,0 +1,79 @@
+
+
+@callback
+def configured_instances(hass):
```
Do we need this if we only allow a single entry?

MartinHjelmare commented on homeassistant/components/abode/__init__.py:
```diff
@@ -131,17 +129,50 @@ def is_light(self, device):
+                CONF_PASSWORD: conf.get(CONF_PASSWORD),
+                CONF_NAME: conf.get(CONF_NAME),
+                CONF_POLLING: conf.get(CONF_POLLING),
+                CONF_EXCLUDE: conf.get(CONF_EXCLUDE),
```
We don't allow excluding specific devices/entities in the config anymore. We can instead use entity disabled options. If possible we should set a default way and include all entities according to that default.
https://github.com/home-assistant/architecture/blob/master/adr/0003-monitor-condition-and-data-selectors.md
https://developers.home-assistant.io/docs/en/entity_registry_disabled_by.html

shred86 replied:
By default, it appears all entities are included which I think is fine since users can then disable entities from the UI (entities page). Would this suffice just relying on the defaults for now? Ideally, I think it would be nice to allow users to select which devices they want to disable during the config entry process, but I'm not quite sure how to do that. The options config flow seems to just be an options dialog for the entire component. I'm not sure if it's possible to have configurable options for each individual device within an integration?

MartinHjelmare replied:
Yes, I think it's enough in this PR to remove these options and let the user disable entitites if needed afterwards.

We don't want options for each configurable device in the options flow. If we can group devices somehow to disable that could be an option. Individual entities are disabled via the options in entity registry.

shred86 replied:
After removing code related to exclude and lights, everything works fine but one unexpected outcome is all switches from Abode show up as one device in HA but two entities, one switch and one light (even if it's not necessarily a light). A user can simply disable the one they don't need, but I'd like to better understand why this is occurring. Will keep looking into it.

As a side note, this will be a "major" change for current users of the Abode component as I'm sure many users have excluded devices in their config file.

MartinHjelmare commented on homeassistant/components/abode/__init__.py:
```diff
@@ -131,17 +129,50 @@ def is_light(self, device):
+                CONF_NAME: conf.get(CONF_NAME),
+                CONF_POLLING: conf.get(CONF_POLLING),
+                CONF_EXCLUDE: conf.get(CONF_EXCLUDE),
+                CONF_LIGHTS: conf.get(CONF_LIGHTS),
```
See above.

shred86 commented:
@MartinHjelmare Thanks for review. Most of these make sense but there's a few I'll have to do some digging around to figure out. Thought I understood what `CONFIG_SCHEMA` was doing but apparently not. Once I get these changes/fixes implemented, should I just submit another push from my repo? I noticed there's an option on here to "commit suggestion" but I don't want my repo to get out of sync.

MartinHjelmare commented:
Regarding suggestions, you can do either. If you commit the suggestion directly in github, you should pull from remote before you do local changes and push again.

shred86 commented on homeassistant/components/abode/__init__.py:
```diff
@@ -131,17 +129,50 @@ def is_light(self, device):
+        lights = []
 
     try:
         cache = hass.config.path(DEFAULT_CACHEDB)
```
I'm struggling with this one. I thought I could simply utilize `async_add_executor_job` when the `AbodeSystem` instance is being created here:

`hass.data[DOMAIN] = AbodeSystem(username, password, cache, name, polling)`

This would turn into:

```
hass.data[DOMAIN] = hass.async_add_executor_job(
            AbodeSystem, username, password, cache, name, polling
        )
```
However, this doesn't appear to be working as the `AbodeSystem` instance isn't  being created.

I'm sure I'm messing up something fairly simple here.

shred86 replied:
Deleted my previous message. I think I understand this now. I basically need to call `hass.async_add_executor_job` when defining `self.abode` in the `AbodeSystem` class.

shred86 replied:
When I try using:

```
hass.data[DOMAIN] = hass.async_add_executor_job(
            AbodeSystem, username, password, cache, name, polling
        )
```

It appears `hass.data[DOMAIN]` is not being setup properly causing errors in every other method or module that references it.

MartinHjelmare replied:
We need to await the task to get the result.
```py
await hass.async_add_executor_job(...)
```

shred86 replied:
So that was the first thing I tried yesterday and I could have sworn it didn't work... of course when I try it again now after you mention it, it seems to be working fine. That was about two hours of my life... lol. Thanks again

shred86 commented on homeassistant/components/abode/config_flow.py:
```diff
@@ -0,0 +1,79 @@
+        password = user_input[CONF_PASSWORD]
+
+        try:
+            Abode(username, password)
```
Not sure how to access the `hass` object from the config_flow.py module since it doesn't appear to be passed into the `AbodeFlowHandler`.

MartinHjelmare replied:
`self.hass`

shred86 replied:
Hm, `async_add_executor_job` doesn't accept kwargs, which is needed for `Abode`.

shred86 commented on homeassistant/components/abode/config_flow.py:
```diff
@@ -0,0 +1,79 @@
+        """Import a config entry from configuration.yaml."""
+        if import_config[CONF_USERNAME] in configured_instances(self.hass):
+            _LOGGER.warning("Account in configuration.yaml is already configured")
+            return await self._show_form({CONF_USERNAME: "identifier_exists"})
```
I'll change it to abort in this case to give the already configured account "priority". I've also changed the logger to "Only one configuration of abode is allowed.". I'm assuming if a user already configured abode then attempts to add it via the configuration.yaml, that log message should be pretty self critiquing.

shred86 committed:
Several fixes from code review

shred86 commented:
Just made a new commit with fixes to all the suggestions. Marked everything as resolved that I verified was corrected.

MartinHjelmare commented on homeassistant/components/abode/config_flow.py:
```diff
@@ -0,0 +1,66 @@
+        try:
+            self.hass.async_add_executor_job(Abode, username, password, True)
+
+        # Need to add error checking if Abode server is down/not responding
```
This remains to do.

shred86 replied:
I think this will require updates to the abodepy module. Right now, this is how `response.status_code` other than 200 is handled:

```
if response.status_code != 200:
            raise AbodeAuthenticationException((response.status_code,
                                                response_object['message']))
```

All that we get back in Home Assistant is a class object, `<class 'abodepy.exceptions.AbodeAuthenticationException'>`.

If I'm understanding this correctly, we would need abodepy to throw a different type of exception for the server being down that we could then use in the Abode component in HA to create another type of error to show the user.

shred86 replied:
So it appears abodepy does pass the error code and error message in the exception, so I should be able to differentiate between invalid credentials (which is being passed as response code 400) and the server not responding, which I need to figure out.

shred86 replied:
I'm not sure how to figure out what the response code is for the abode server being down. I'm sure there's a few different cases we would have to cover. The only thing I can really is simulate is your own ISP being down which HA already throws an error for. However, I did make some adjustments to make the error catching more explicit for the credentials being invalid, then sort of a catch all for any other errors on the Abode server side called "abode_error" with a message "Error occurred with Abode servers."
```
        except AbodeAuthenticationException as error:
            if error.errcode == 400:
                return await self._show_form({"base": "invalid_credentials"})
            return await self._show_form({"base": "abode_error"})
```
I think this implementation covers what we need and at least provides the user with more descriptive feedback.

MartinHjelmare commented on tests/components/abode/test_config_flow.py:
```diff
@@ -0,0 +1,77 @@
+    flow = config_flow.AbodeFlowHandler()
+    flow.hass = hass
+
+    with patch.object(hass.config_entries, "async_entries"):
```
Add a mock entry instead.

https://github.com/home-assistant/home-assistant/blob/71a351605382136c5580a2eb8f46d5b24216ac64/tests/components/hue/test_config_flow.py#L68

MartinHjelmare commented on tests/components/abode/test_config_flow.py:
```diff
@@ -0,0 +1,77 @@
+
+async def test_invalid_credentials(hass):
+    """Test that invalid credentials throws an error."""
+    from abodepy.exceptions import AbodeAuthenticationException
```
Move 3rd party imports to the top of the module.

MartinHjelmare commented on homeassistant/components/abode/switch.py:
```diff
@@ -3,39 +3,35 @@
+
+async def async_setup_entry(hass, config_entry, async_add_devices):
     """Set up Abode switch devices."""
     import abodepy.helpers.constants as CONST
```
Please move 3rd party imports to the top of the module.

MartinHjelmare commented on homeassistant/components/abode/__init__.py:
```diff
@@ -110,69 +106,72 @@ def __init__(self, username, password, cache, name, polling, exclude, lights):
+    """Set up Abode component via config entry (Integrations UI)."""
+    username = config_entry.data.get(CONF_USERNAME)
+    password = config_entry.data.get(CONF_PASSWORD)
+    name = config_entry.data.get(CONF_NAME)
```
We aren't saving the extra options available in the config yaml in the config entry data. We need to do that if they should be available here.

MartinHjelmare replied:
Where do we use the name option?

shred86 replied:
I’m not quite sure actually. A lot of this is obviously from whoever created this component previously. I just assumed the name option was an override for a property that Home Assistant used. I personally don’t see a need for it and think it’s simpler using the default which is the component name, especially since we can only have one instance right now.

MartinHjelmare commented on homeassistant/components/abode/__init__.py:
```diff
@@ -96,7 +92,7 @@
-    def __init__(self, username, password, cache, name, polling, exclude, lights):
+    def __init__(self, username, password, cache, name, polling):
         """Initialize the system."""
         import abodepy
```
Please create the Abode instance before passing it to the AbodeSystem class. We want to avoid side effects in init method.

MartinHjelmare commented on tests/components/abode/test_config_flow.py:
```diff
@@ -0,0 +1,77 @@
+
+async def test_step_import(hass):
+    """Test that the import step works."""
+    conf = {CONF_USERNAME: "user@email.com", CONF_PASSWORD: "password"}
```
We should try passing all the options to make sure that works.

MartinHjelmare commented:
Please solve the merge conflict.

shred86 commented on homeassistant/components/abode/config_flow.py:
```diff
@@ -0,0 +1,66 @@
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class AbodeFlowHandler(config_entries.ConfigFlow):
```
@ludeeus The blueprint config_flow.py doesn’t match the developer documents. Not sure if that’s intentional or perhaps something that needs to be updated as well.

shred86 commented on homeassistant/components/abode/config_flow.py:
```diff
@@ -0,0 +1,66 @@
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        from abodepy import Abode
```
Is there any case where 3rd party imports should occur at the start of a method? This is how it was previously so I left it. I've just noticed a couple other methods that have 3rd party imports, such as `setup_abode_events`.

MartinHjelmare replied:
Please move all of them. After we introduced manifest.json we can now have all 3rd party imports at the top of the module.

shred86 committed:
Several fixes from second code review

shred86 commented:
Another commit with fixes to all the suggestions. Also realized the unit test wasn't actually doing what I thought it was - should be fixed now and added another test case for the updated exception handling.

I'm not sure why there is a branch merge conflict. I haven't touched the manifest.json file but it looks like there's a mismatch between `codeowners`. The last time this branch merge conflict occurred, it resolved itself.

MartinHjelmare commented:
I don't think the merge conflict will resolve itself.

MartinHjelmare commented on homeassistant/components/abode/__init__.py:
```diff
@@ -96,83 +94,84 @@
         cache = hass.config.path(DEFAULT_CACHEDB)
-        hass.data[DOMAIN] = AbodeSystem(
-            username, password, cache, name, polling, exclude, lights
+        abode = Abode(
```
Now we need to schedule this instantiation on the executor thread pool.

MartinHjelmare commented on homeassistant/components/abode/__init__.py:
```diff
@@ -96,83 +94,84 @@
+            get_automations=True,
+            cache_path=cache,
+        )
+        hass.data[DOMAIN] = await hass.async_add_executor_job(
```
We don't need the executor here anymore.

MartinHjelmare commented on homeassistant/components/abode/__init__.py:
```diff
@@ -96,83 +94,84 @@
     """Abode System class."""
 
-    def __init__(self, username, password, cache, name, polling, exclude, lights):
+    def __init__(self, abode, username, password, cache, polling):
```
We don't need the credentials and cache here anymore.

MartinHjelmare commented on homeassistant/components/abode/__init__.py:
```diff
@@ -96,83 +94,84 @@
 
     for platform in ABODE_PLATFORMS:
-        discovery.load_platform(hass, platform, DOMAIN, {}, config)
+        await hass.config_entries.async_forward_entry_unload(config_entry, platform)
```
We can use asyncio.wait or asyncio.gather to await all coroutines concurrently.

shred86 replied:
This is new to me (much like the rest of Python) but this is what I gather after doing a bit of research and just wanted to make sure I have this correct:
```
from asyncio import gather
````
Then:

```
    for platform in ABODE_PLATFORMS:
        await gather(
            hass.config_entries.async_forward_entry_unload(config_entry, platform)
        )
```

MartinHjelmare replied:
We need to collect the coroutine objects in a list first. Then use gather on the items in the list.

shred86 replied:
Ah makes sense - I had a feeling that was the case reading some of the docs. Something like this:

```
    tasks = []

    for platform in ABODE_PLATFORMS:
        tasks.append(
            hass.config_entries.async_forward_entry_unload(config_entry, platform)
        )

    await gather(*tasks)
```

MartinHjelmare commented on homeassistant/components/abode/binary_sensor.py:
```diff
@@ -1,19 +1,24 @@
+
 
-    data = hass.data[ABODE_DOMAIN]
+async def async_setup_entry(hass, config_entry, async_add_devices):
```
Please rename `async_add_devices` to `async_add_entities` in all platforms.

MartinHjelmare commented on homeassistant/components/abode/config_flow.py:
```diff
@@ -0,0 +1,66 @@
+            data={CONF_USERNAME: username, CONF_PASSWORD: password},
+        )
+
+    async def _show_form(self, errors=None):
```
We don't use `await` inside so this doesn't need to be a coroutine. Decorate it with `@callback` instead.

MartinHjelmare commented on homeassistant/components/abode/config_flow.py:
```diff
@@ -0,0 +1,66 @@
+
+    def __init__(self):
+        """Initialize."""
+        self.data_schema = {}
```
We could make this a `Schema` instance directly.

shred86 replied:
I may have messed this one up and will need to fix. I updated to: 
```
        self.data_schema = vol.Schema(
            {vol.Required(CONF_USERNAME): str, vol.Required(CONF_PASSWORD): str}
        )
```
Thought it worked when I tried it but it's throwing `TypeError: unhashable type: 'Schema'` now.

shred86 replied:
Fixed. Will be in the next commit:
```
        self.data_schema = {
            vol.Required(CONF_USERNAME): str,
            vol.Required(CONF_PASSWORD): str,
        }
```

shred86 committed:
Several fixes from third code review

shred86 committed:
Update documentation url to fix branch conflict

shred86 commented:
Another commit to fix the code review suggestions. I also updated the manifest.json - I have no idea why it's flagging a branch conflict for me adding:

`"config_flow": true`

shred86 reviewed (commented):
Forgot to delete this - corrected on the next commit.

Edit: Sorry, thought it would reference the line. It's line 67 in `__init__.py`: `vol.Optional(CONF_NAME): cv.string,`

And line 21: `CONF_NAME`, line 35: `NOTIFICATION_ID = "abode_notification"` and line 36 `NOTIFICATION_ID = "abode_notification"`.

None of these constants are  being used and will be removed:
```
EVENT_ABODE_ALARM = "abode_alarm"
EVENT_ABODE_ALARM_END = "abode_alarm_end"
EVENT_ABODE_AUTOMATION = "abode_automation"
EVENT_ABODE_FAULT = "abode_panel_fault"
EVENT_ABODE_RESTORE = "abode_panel_restore"
```

MartinHjelmare commented:
If the dev branch changes the same file as we change here and git can't resolve, there will be a merge conflict. The manifest.json file on dev branch was changed three days ago. We need to solve that merge conflict by rebasing on latest dev branch and specify how to solve the conflict.

shred86 committed:
Fixes config flow and removes unused constants

shred86 commented:
Rebased and committed some small changes I mentioned above. Additionally, updated the original post in this PR with the break changing and updated the PR for the documentation. I've also fixed [the issue](https://github.com/home-assistant/home-assistant/pull/26699#discussion_r330839234) with switches being added as switches and lights from Abode which I'll include in the next commit (will wait a bit to allow for review) along with some minor reorganizing of the imports to adhere to the [style guidelines](https://www.python.org/dev/peps/pep-0008/#imports).

shred86 committed:
Fix for switches, polling entry option, improved tests

shred86 commented:
Added option to enable polling to config flow UI. Personally, I think we can just get rid of polling since cloud push is obviously much faster and less demand on abode's servers. I added the label "Polling (not recommended)" to make it fairly clear it shouldn't be used normally. I think a better option might be to remove the option from the UI and only allow enabling polling through the configuration.yaml. However, I tried reworking config_flow.py but ran into issues with getting exceptions to assert properly when moving the credential testing to a separate function. I'll continue trying to work.

Additionally, pylint is complaining about `from .const import DOMAIN` in `config_flow.py` not being used when it is in fact being used:
```
class AbodeFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
    """Config flow for Abode."""
```
Not entirely sure how to clear that up as it's failing CI tests.

MartinHjelmare commented on homeassistant/components/abode/config_flow.py:
```diff
@@ -24,6 +26,7 @@ def __init__(self):
         self.data_schema = {
             vol.Required(CONF_USERNAME): str,
             vol.Required(CONF_PASSWORD): str,
+            vol.Optional(CONF_POLLING, default=False): bool,
```
This option should be part of the options flow, since it's optional and isn't needed to set up the connection correctly.

https://developers.home-assistant.io/docs/en/config_entries_options_flow_handler.html

MartinHjelmare replied:
We can do this in a follow up PR. For now we can just keep the option in config yaml.

shred86 replied:
The issue I ran into is when using the configuration.yaml file and `polling` is set to `True`, we run into an error with config_flow.py. My understanding is from config_flow.py, the `async_step_import` method handles the configuration.yaml data but right now, it's simply being passed to `async_step_user`:

`return await self.async_step_user(import_config)`

Within `async_step_user`, I don't have any variables defined for `polling` since it's not a part of the UI config flow, so it causes an error. I'm pretty sure there's a simple way to do this but I spent a few hours last night and couldn't figure it out. Basically I need a way to default `polling` to `False` unless it's passed in from `async_step_import`.

shred86 replied:
And of course after 5 mins looking at it again, I think I figured out a simple solution. Within the `async_step_user` of the flow handler:

```
        if CONF_POLLING in user_input:
            polling = user_input[CONF_POLLING]
        else:
            polling = False
```

MartinHjelmare commented:
We can disable the pylint warning about non used import of `DOMAIN` in the config flow. It's a bug in pylint.

MartinHjelmare reviewed (commented):
Make sure to not exclude the config_flow.py module from coverage calculation.

https://github.com/home-assistant/home-assistant/blob/2c6a869bc61d922b77cc4a7588d079e8a2836013/.coveragerc#L13

shred86 committed:
Update .coveragerc, disable pylint W0611, remove polling from UI

shred86 commented:
New push with the mentioned fixes.

Looking at the structure of `ConfigEntry` in [config_entries.py](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/config_entries.py), there is a `data` dict where we're currently saving `username`, `password` and `polling`, as well as an `optional` dict for optional configuration options. Since I plan on adding the `polling` option to the options config flow in another PR, would it be better to setup `config_flow.py` to save the `polling` setting within the `options` dict now? I realize it's fairly minor, but it would just prevent `ConfigEntry` from having an unused `polling` key after the options config flow is implemented.

MartinHjelmare commented on homeassistant/components/abode/config_flow.py:
```diff
@@ -0,0 +1,76 @@
+
+        username = user_input[CONF_USERNAME]
+        password = user_input[CONF_PASSWORD]
+        polling = user_input[CONF_POLLING] if CONF_POLLING in user_input else False
```
```py
polling = user_input.get(CONF_POLLING, False)
```

shred86 replied:
That is a lot simpler.

MartinHjelmare reviewed (commented):
To save an option in the options dict in the config entry we need an options flow.

We could store the option in hass.data temporarily until we implement the options flow.

shred86 commented:
> To save an option in the options dict in the config entry we need an options flow.
> 
> We could store the option in hass.data temporarily until we implement the options flow.

Ah, okay. That sounds good - I don't want to grow this PR anymore than it already has. Once this is finalized, I'll start working on the options flow.

shred86 commented:
One minor clean up item I'm not sure how to handle. When a user removes the integration from the UI then shuts down HA, the following error asserts:
```
^C2019-10-06 10:48:50 ERROR (MainThread) [homeassistant.core] Error doing job: Future exception was never retrieved
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/myuser/.homeassistant/custom_components/abode/__init__.py", line 220, in logout
    if not hass.data[DOMAIN].polling:
KeyError: 'abode'
```
The reason being is when abode is setup, the `setup_hass_events` function is executed which calls `hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, logout)`. This is to execute the `logout` method in `abodepy.Abode` which is necessary. However, when a user deletes the abode integration and the `async_unload_entry` function is ran, it executes `hass.data.pop(DOMAIN)`, thus the abode domain no longer exists which is causing that error.

Is there some sort of way to "undo" the `hass.bus.async_listen_once` when `async_unload_entry` is ran?

MartinHjelmare commented:
All listeners that the integration creates must be unsubscribed in `async_unload_entry`. Save the return value when subscribing to an event. The return value is a function to unsubscribe the listener.

MartinHjelmare commented:
The update method will be scheduled in the executor thread pool. So that should be fine. How long is the usual response time of the device when calling refresh?

shred86 commented:
> All listeners that the integration creates must be unsubscribed in `async_unload_entry`. Save the return value when subscribing to an event. The return value is a function to unsubscribe the listener.

Ah okay. I'm not sure if this is the most elegant solution so I'll post it here for review before I make the commit. Within `setup_hass_events` I have:
```
hass.data[DOMAIN].logout_listener = hass.bus.async_listen_once(
        EVENT_HOMEASSISTANT_STOP, logout
    )
```
I'm assuming `hass.data[DOMAIN]` is the best place to store objects. From there, I'm simply executing `hass.data[DOMAIN].logout_listener()` in the `async_unload_entry` function.

> The update method will be scheduled in the executor thread pool. So that should be fine. How long is the usual response time of the device when calling refresh?

I'm not sure on the exact response time but just based on what I can see from the UI, it seems very fast (near real time).

shred86 commented on homeassistant/components/abode/config_flow.py:
```diff
@@ -0,0 +1,76 @@
+
+        except AbodeAuthenticationException as error:
+            if error.errcode == 400:
+                return self._show_form({"base": "invalid_credentials"})
```
When abode is configured using the configuration.yaml file, if the credentials are invalid, there is currently no way for the user to know. What shows up in the UI is abode is discovered and the user can click "Configure" to setup abode via the config flow.

I think adding some logging here that matches the config flow messages would be useful:

`_LOGGER.warning("Invalid credentials.")`

`_LOGGER.warning("Error occurred with Abode server.")`

shred86 commented:
Just made a [pull request](https://github.com/MisterWil/abodepy/pull/49) over on abodepy. Each device in the Abode system has a `device_id` (ex: RF:0171bd10) and a `device_uuid` (ex: 6d54a1fabd662015ae8d2b2c72d65df6). Currently, I have the Abode component using the `device_id` for entity registry. However, I think the `device_uuid` from Abode would be better to use as it's far more unique. I've noticed that Z-Wave devices paired to Abode will be assigned a `device_id` of something like ZW:00000001, which is obviously not as unique.

I've @MisterWil a message in hopes that he can merge the PR and bump the abodepy version in PyPI. Once that's done, I'll make a commit with the change mentioned above.

shred86 committed:
Multiple fixes and edits to adhere to style guidelines

shred86 commented:
New commit that resolves most of the issues mentioned above. Thanks to @MisterWil for making a quick turn and getting my abodepy PR merged. Will add some comments in the changes.

shred86 commented on homeassistant/components/abode/alarm_control_panel.py:
```diff
@@ -74,3 +73,8 @@ def device_state_attributes(self):
+    @property
+    def unique_id(self):
+        """Return a unique ID to use for this device."""
+        return f"{self._device.name} {self._device.device_id}"
```
The alarm control panel does not have a `uuid` in Abode, only the devices paired to it. I looked through all of the data that is passed in the JSON for the alarm control panel and there really isn't a good option like `uuid`. The `id` assigned to the alarm control panel is `area_1`, which isn't very unique so I added the `name` property which gives it a `unique_id` in Home Assistant of: `Abode Alarm area_1`.

It's not ideal as I'd prefer to have something closer to `uuid` like the rest of the devices have. However, this should suffice in terms of being unique to any other integrations that might be added to Home Assistant (I can't imagine anything outside of Abode being assigned `Abode Alarm area_1`.

That being said, if we ever add support to have multiple Abode integrations into a single HA installation, then this `unique_id` would not work as they would conflict with each other.

MartinHjelmare replied:
The unique_id only needs to be unique within the integration domain and the platform domain. So it needs to differentiate between abode alarm_control_panel entities.

We shouldn't add a unique id here if there is no unique stable identifier.

shred86 replied:
Ah, okay. It just seems odd not having a unique id since every other device will show up in the entity registry except the alarm control panel.

shred86 replied:
Figured out a solution. The mac address of the Abode hub (alarm) is contained within the login JSON data. I've submitted a [pull request](https://github.com/MisterWil/abodepy/pull/50) for abodepy that allows the mac address to be accessed from the `alarm` module. I spoke with @MisterWil and he will hopefully be able to merge the PR tomorrow and bump the abodepy version. Once that's done, I'll commit a small addition to this `alarm_control_panel.py` module:

```
    @property
    def unique_id(self):
        """Return a unique ID to use for this device."""
        return self._device.mac_address
```

Hopefully that should be it!

shred86 replied:
Okay, @MisterWil did some great work such that the abode alarm (hub) will return its mac address when the same `self._device.device_uuid` is called so overriding `def unique_id(self):` in the alarm_control_panel.py is not necessary. Just pushed a commit that bumps abodepy to 0.16.3 which includes these changes.

shred86 commented on homeassistant/components/abode/alarm_control_panel.py:
```diff
@@ -27,8 +27,7 @@
     async_add_entities(
-        [AbodeAlarm(data, await hass.async_add_executor_job(data.abode.get_alarm))],
-        True,
+        [AbodeAlarm(data, await hass.async_add_executor_job(data.abode.get_alarm))]
```
Removed `True` to resolve the issue with the alarm control panel not showing up in the integrations UI after adding the abode integration. I'm not sure what's causing the issue, but I'm also not sure why HA needs to refresh the device when it's added (as I understand that's why the `True` argument does). The reason being is this is occurring at start up or when the Abode integration is added, which the device states would be updated just seconds ago.

shred86 committed:
Removed unique_id

shred86 committed:
Minor correction for formatting error in rebase

shred86 commented:
This PR will resolve #27284.

shred86 committed:
Resolves issue causing CI to fail

shred86 committed:
Bump abodepy version

shred86 commented:
I disabled a few entities from the entity registry UI and noticed a warning in the logs with the camera:

`2019-10-10 18:11:33 WARNING (MainThread) [homeassistant.helpers.entity] Entity camera.living_room_camera is incorrectly being triggered for updates while it is disabled. This is a bug in the abode integration.`

Only appears to be the camera showing this warning. I'm wondering if it's caused by this method in the `AbodeCamera` class:

```
async def async_added_to_hass(self):
        """Subscribe Abode events."""
        await super().async_added_to_hass()

        self.hass.async_add_job(
            self._data.abode.events.add_timeline_callback,
            self._event,
            self._capture_callback,
        )
```

I'm assuming when an entity is disabled, the camera device is still subscribed to receiving updates.

MartinHjelmare commented:
There's a delay of 30 seconds before the integration is reloaded after an entity is enabled/disabled in the entity registry to avoid the integration reloading multiple times if a user clicks around in the entity registry.

https://github.com/home-assistant/home-assistant/blob/4c71c6df6f584bf220d1f9051ba7df55317c46b1/homeassistant/config_entries.py#L771

https://github.com/home-assistant/home-assistant/blob/4c71c6df6f584bf220d1f9051ba7df55317c46b1/homeassistant/config_entries.py#L823-L831

shred86 commented:
Ah okay, that makes sense. I thought maybe it was an issue with the Abode component. There really should be as all note at a minimum in the UI that explains that. I can only imagine that will be confusing for some users (like me). 😊

Well, I’m hoping my PR for abodepy will be merged tomorrow and updated. I’ll commit the change mentioned above once that’s done and hopefully that’s the last thing...

shred86 committed:
Add remove device callback and minor clean up

MartinHjelmare commented on homeassistant/components/abode/__init__.py:
```diff
@@ -100,73 +84,79 @@
-        """Check if an automation is configured to be excluded."""
-        return automation.automation_id in self.exclude
+async def async_setup(hass, config):
+    """Set up Abode platform."""
```
This is the component module. We could write "integration" which is the whole package.

shred86 replied:
Sounds good. I was just referencing what the Hue platform had for their doc strings. :) Just pushed a commit with it changed to "integration".

MartinHjelmare commented on homeassistant/components/abode/__init__.py:
```diff
@@ -100,73 +84,79 @@
-    exclude = conf.get(CONF_EXCLUDE)
-    lights = conf.get(CONF_LIGHTS)
+async def async_setup_entry(hass, config_entry):
+    """Set up Abode platform from a config entry."""
```
See above.

shred86 committed:
Fix incorrect method name

shred86 committed:
Docstring edits

shred86 committed:
Fix duplicate import issues from rebase

shred86 commented:
Hopefully that should be the last commit. =)

MartinHjelmare commented on homeassistant/components/abode/__init__.py:
```diff
@@ -239,9 +225,11 @@ def logout(event):
+        await hass.async_add_executor_job(hass.data[DOMAIN].abode.events.start)
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, logout)
+    hass.data[DOMAIN].logout_listener = hass.bus.async_listen_once(
```
We should initialize the `logout_listener` attribute in the `AbodeSystem` class. Set the attribute to  `None`.

MartinHjelmare commented on homeassistant/components/abode/__init__.py:
```diff
@@ -100,73 +84,79 @@
-        self.exclude = exclude
-        self.lights = lights
         self.devices = []
 
```
Set `logout_listener` to `None` here.

shred86 committed:
Add logout_listener attribute to AbodeSystem

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare reviewed (changes requested):
Sorry, I noticed that the coverage isn't 100% for the config flow module.

https://codecov.io/gh/home-assistant/home-assistant/compare/a82ff4f7a94fc6f2c635f9ecc260211654fcc42d...9f01395f65fa894e8d88294e6836d3575e29f3d2/diff#D1-77

shred86 committed:
Add additional test for complete coverage

shred86 commented:
Added additional tests to the `test_one_config_allowed` function. Previously it was only covering the case where `async_step_user` was being called with an existing config entry. I simply added onto that same function and added a import config file and called `async_step_import` so both cases are covered now.

MartinHjelmare reviewed (approved):
Good!

shred86 commented:
Thanks for all the help @MartinHjelmare - I learned a ton in this process which I realize comes at the expense of your time fixing all my errors. I started learning python/programming primarily to help contribute to Home Assistant. I'm hoping to continue learning and contributing where I can. Cheers.

shred86 commented:
Noticed one minor issue - there is no `.translations` folder in the abode component folder. I was under the impression the strings.json file was used to create that translations folder, but not something I was supposed to do on my end.
